CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,474 @@
+
+**Linear backoff:** A rate limiting scheme allowing up to a given number of requests in a defined
+time period could be implemented as a linear backoff with some amount of saved-up quota cap.
+
+```typescript
+type LinearBackoffDomain = {
+  name: "ODIS Linear Backoff Domain"
+  version: 1
+  // Length of time, in milliseconds, to refresh a single unit of quota.
+  // Undefined refresh time indicates that the quota never refreshes (hard cap).
+  refresh?: number
+  // Maximum number of saved quota.
+  cap: number
+  // Unique string to identify the "instance" of this rate limited domain.
+  salt?: string
+}
+```
+
+**Not before:** Useful in the lotteries and other random-selection based protocols, a domain could
+be created to restrict requests before a particular time, but publicly available after.
+
+```typescript
+type NotBeforeDomain = {
+  name: "ODIS Not Before Domain"
+  version: 1
+  // Unix time after which this domain can be queried.
+  notBefore: number
+}
+```
+
+**Peppered:** Useful in the context of hardening encryption keys for non-public ciphertexts, a
+domain could be defined that requires a signature from a pepper-derived key-pair. The pepper may be
+stored alongside the ciphertext.
+
+```typescript
+type PepperedDomain = {
+  name: "ODIS Peppered Domain"
+  version: 1
+  // Public key of a key-pair derived from the pepper.
+  publicKey: string
+  // Maximum number of requests that can be made against this domain.
+  cap?: number
+}
+
+type PepperedDomainOptions = {
+  // Nonce value to include in the signature message to make it unique.
+  nonce: number
+  // EIP-712 signature over the request.
+  signature: string
+}
+```
+
+**Phone number hashing:** The current ODIS rate limiting scheme could be implemented as a particular
+domain in the new API. A domain string and options could be written as follows:
+
+```typescript
+type PhoneNumberDomain = {
+  name: "ODIS Phone Number Domain"
+  version: 1
+}
+
+type PhoneNumberDomainOptions = {
+  // Celo account address. Query is charge against this account's quota.
+  account: string
+  // Signature from the account, or authorized signer for the account.
+  authorization: string
+  // Optional on-chain identifier. Unlocks additional quota if the
+  // account is verified as an owner of the identifier.
+  hashedPhoneNumber?: string
+}
+```
+
+**Smart contract access controlled:** Useful for expanding the set of use cases to community derived
+ones. The address of a smart contract implementing the following interface will be a parameter in
+the domain.
+
+```solidity
+interface OdisAccessController {
+  function shouldAllowOdisRequest(bytes calldata domainData, bytes calldata authorization) external view returns(bool);
+}
+```
+
+For example, this contract could check that a user owns a certain NFT or has burned a fee before
+being allowed access to the signature. The signature would be applied to the domain and the returned
+value from the contract function.
```

Comment:
Not something we need to worry about now but we should think about how we could ensure backwards compatibility of the smart contract logic (i.e. prevent the logic from changing without a version increase)

Reply:
That's a good callout. If we decide to implement this domain, we'd need to deal with issues like that.